vendor/bat-native-ads/src/bat/ads/internal/ads/serving/eligible_ads/pipelines/notification_ads/eligible_notification_ads_v3.cc

```diff
@@ -0,0 +1,122 @@
+/* Copyright (c) 2021 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "bat/ads/internal/ads/serving/eligible_ads/pipelines/notification_ads/eligible_notification_ads_v3.h"
+
+#include "absl/types/optional.h"
+#include "base/bind.h"
+#include "bat/ads/internal/ads/ad_events/ad_events_database_table.h"
+#include "bat/ads/internal/ads/serving/choose/predict_ad_embeddings.h"
+#include "bat/ads/internal/ads/serving/eligible_ads/exclusion_rules/exclusion_rules_util.h"
+#include "bat/ads/internal/ads/serving/eligible_ads/exclusion_rules/notification_ads/notification_ad_exclusion_rules.h"
+#include "bat/ads/internal/ads/serving/serving_features.h"
+#include "bat/ads/internal/ads/serving/targeting/user_model_info.h"
+#include "bat/ads/internal/ads_client_helper.h"
+#include "bat/ads/internal/base/logging_util.h"
+#include "bat/ads/internal/creatives/notification_ads/creative_notification_ads_database_table.h"
+#include "bat/ads/internal/geographic/subdivision/subdivision_targeting.h"
+#include "bat/ads/internal/resources/behavioral/anti_targeting/anti_targeting_resource.h"
+#include "bat/ads/internal/segments/segment_alias.h"
+
+namespace ads::notification_ads {
+
+EligibleAdsV3::EligibleAdsV3(
+    geographic::SubdivisionTargeting* subdivision_targeting,
+    resource::AntiTargeting* anti_targeting_resource)
+    : EligibleAdsBase(subdivision_targeting, anti_targeting_resource) {}
+
+void EligibleAdsV3::GetForUserModel(
+    const targeting::UserModelInfo& user_model,
+    GetEligibleAdsCallback<CreativeNotificationAdList> callback) {
+  BLOG(1, "Get eligible notification ads");
+
+  database::table::AdEvents database_table;
+  database_table.GetForType(
+      mojom::AdType::kNotificationAd,
+      [=](const bool success, const AdEventList& ad_events) {
+        if (!success) {
+          BLOG(1, "Failed to get ad events");
+          callback(/*had_opportunity*/ false, {});
+          return;
+        }
+
+        GetBrowsingHistory(user_model, ad_events, callback);
+      });
+}
+
+///////////////////////////////////////////////////////////////////////////////
+
+void EligibleAdsV3::GetBrowsingHistory(
+    const targeting::UserModelInfo& user_model,
+    const AdEventList& ad_events,
+    GetEligibleAdsCallback<CreativeNotificationAdList> callback) {
+  const int max_count = features::GetBrowsingHistoryMaxCount();
+  const int days_ago = features::GetBrowsingHistoryDaysAgo();
+  AdsClientHelper::GetInstance()->GetBrowsingHistory(
+      max_count, days_ago,
+      base::BindOnce(&EligibleAdsV3::GetEligibleAds, base::Unretained(this),
+                     user_model, ad_events, callback));
+}
+
+void EligibleAdsV3::GetEligibleAds(
```

Comment:
Some of the previous logic with getting the `user_model` can probably be adjusted/removed, since we don't use it for v3 serving. And since we are retrieving the browsing history we need in `PredictAdEmbeddings` during `GetTextEmbeddingHtmlEventsFromDatabase`, can we call `GetEligibleAds` right after we get the `AdEventList` as in line 38?

Reply:
removed `user_model` from the call hierarchy. As discussed, `GetBrowserHistory` is still needed for filtering creatives